main/src/ca/uwaterloo/flix/language/phase/Kinder.scala

```diff
@@ -1022,6 +1031,10 @@ object Kinder {
       Validation.success(KindedAst.Expr.Error(m, tvar, evar))
   }
 
+  private def visitDefn(defn0: ResolvedAst.Expr.Def)(implicit scope: Scope, flix: Flix): Validation[KindedAst.Expr.Def, KindError] = defn0 match {
```

Comment:
Why a separate function here?

Reply:
If we want to avoid a type cast in `visitExp` in the `ApplyDef` case then we need a function that returns a `KindedAst.Expr.Def` so we can construct a `KindedAst.Expr.ApplyDef` which takes not just an `Expr` but a `Expr.Def`